### ISSUE TITLE
[BUG] Instance manager pod `awsIAMRoleArn` annotation disappearing

### ISSUE THREAD

**Issue by NRCan-LGariepy** [longhorn/longhorn#9923]
## Describe the bug

After applying an upgrade to Longhorn via the Rancher UI, the annotations that are required to interact with the S3 backup target via kube2IAM are no longer present on the new instance-managers, which makes backup jobs fail.

Our cluster is managed by Rancher, provided by AWS EC2 (not EKS).

## To Reproduce

Set up a backup target as outlined in the [docs](https://longhorn.io/docs/1.7.2/snapshots-and-backups/backup-and-restore/set-backup-target/#set-up-aws-s3-backupstore), using Kube2IAM.

Then, upgrade Longhorn to a new version (e.g. 1.71 -> 1.72). It's still unclear to us what exactly triggers this behavior, we have also seen it happen when nodes are reprovisionned/recreated, for example when upgrading RKE2.

The new instance managers do not have the role annotation required to interact with the backup target. This causes recurring backup jobs to fail.

## Expected behavior

I expect the new instance managers to also have the required role annotation after an upgrade.

## Environment

 - Longhorn version: 1.70, 1.71, 1.72 (at least those)
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): Rancher catalog
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: RKE2 1.29, HA control plane (3 nodes) with 3 worker nodes having access to Longhorn.
 - Node config : Ubuntu 24.04 
   - OS type and version:
   - Kernel version:
   - CPU per node:
   - Memory per node:
   - Disk type (e.g. SSD/NVMe/HDD):
   - Network bandwidth between the nodes (Gbps):
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): AWS EC2
 - Number of Longhorn volumes in the cluster: 13

## Workaround and Mitigation

It seems that after running the upgrade and observing the missing annotations, making any change to the longhorn app yaml config reapplies the annotations.

**Comment by derekbit** [comment]
cc @ChanYiLin @mantissahz

**Comment by mantissahz** [comment]
Hi @NRCan-LGariepy,

Could you provide the support bundle for the investigation?

**Comment by NRCan-LGariepy** [comment]
> Hi @NRCan-LGariepy,
> 
> Could you provide the support bundle for the investigation?

Is there a way to share the bundle privately/securely? I don't know what the bundle contains exactly but I need to ensure I don't share sensitive information.

**Comment by NRCan-LGariepy** [comment]
@mantissahz Thank you. Sent.

For some added context, the way we noticed this problem initially was that our longhorn backup jobs would start failing for seemingly no reason. After some troubleshooting we think this happens only after bumping longhorn's version.

**Comment by NRCan-LGariepy** [comment]
> [@NRCan-LGariepy](https://github.com/NRCan-LGariepy),
> 
> Could you mail the bundle to the [longhorn-support-bundle@suse.com](mailto:longhorn-support-bundle@suse.com)?

@mantissahz I have finally got around to sending the support bundle. Our corporate email server prevents sending large attachments, so I included a dropbox link and a password that is required to unzip the archive.

Please let me know if there's anything else I can do to help, this bug is really causing us some problems!

Thank you.

**Comment by NRCan-LGariepy** [comment]
We haven't hit this issue for a while. I will close this issue for now, will reopen if I have more information to share. In the meantime, feel free to reopen if you think there's still work that can be done on this.

**Comment by NRCan-LGariepy** [comment]
Seems like we have once again hit this issue, backups would not go through, culprit was once again instance managers were missing their AWS role annotation.

**Comment by mantissahz** [comment]
Hi @NRCan-LGariepy,
Could you please provide the new support bundle and indicate which instance managers encountered this issue?

**Comment by NRCan-LGariepy** [comment]
Hi @mantissahz thanks for your help. Before we move on to the support bundle, as a sanity check, could you clarify exactly which Longhorn components require the Kube2IAM annotation (i.e. which one interact with AWS/S3), and what way does Longhorn provide to apply annotations to those pods (either via the chart or other).

I read the docs of course, but it's not super clear on these two particular points.

Thanks!

**Comment by mantissahz** [comment]
>could you clarify exactly which Longhorn components require the Kube2IAM annotation (i.e. which one interact with AWS/S3)

The pod is created by the Longhorn manager using the Longhorn instance manager image, and it involves the Longhorn engines when creating a backup.

>what way does Longhorn provide to apply annotations to those pods (either via the chart or other)

The kubernetes secret controller of the longhorn-manager will apply the annotations to the instance manager pods if the `AWS_IAM_ROLE_ARN` in the secret is updated.

**Comment by NRCan-LGariepy** [comment]
Great, thank you.

Maybe I'm completely wrong, but surely there are instances where the secret is not updated but the annotations still need to be re-applied, no? For example node rotation.

**Comment by mantissahz** [comment]
>but surely there are instances where the secret is not updated but the annotations still need to be re-applied, no? For example node rotation.

What kind of instances did you mean? And what did you mean that annotations still need to be reapplied for node rotation? Could you create a scenario to describe it?

**Comment by NRCan-LGariepy** [comment]
Sure, let's say a new node with longhorn storage is added to the cluster, then longhorn needs to spawn a new instance manager (among other things). That instance manager should have the annotation to work properly. So, if I understand correctly, that is a situation where the annotations needs to be applied even though the secret has not been updated.

I hope that makes sense, let me know if not. Thanks.

**Comment by NRCan-LGariepy** [comment]
@mantissahz Yesterday I upgraded longhorn to 1.9.2, along with the engine images. As a result, I now have duplicate instance managers (for 1.9.1 and 1.9.2). The 1.9.1 ones do have the aws annotation, while the new 1.9.2 instance managers do not.

Here are some screenshots.

<img width="1506" height="340" alt="Image" src="https://github.com/user-attachments/assets/7c28d08e-3e46-4692-a7e2-a85b47d07706" />
<img width="1522" height="365" alt="Image" src="https://github.com/user-attachments/assets/384ca4c4-3a5a-4773-8762-4e7e1c18d6e2" />
<img width="1525" height="528" alt="Image" src="https://github.com/user-attachments/assets/73eaa8d7-5517-425f-a63d-136c0319d29a" />

This is really making it difficult for us to work with Longhorn as when this happens, backups start failing.

**Comment by mantissahz** [comment]
Hi @NRCan-LGariepy,
Could you provide the new support bundle and indicate which IM pods are problematic for the investigation?

**Comment by NRCan-LGariepy** [comment]
@mantissahz I sent a link to the support bundle over at longhorn-support-bundle@suse.com. Please let me know if there's anything else I can do.

**Comment by derekbit** [comment]
@mantissahz Can we simply apply the annotation to the newly created IM?

**Comment by NRCan-LGariepy** [comment]
Hey, is there anything I can do to help this issue move forward? Thanks.

**Comment by COLDTURNIP** [comment]
The Kubernetes secret controller does attach IAM role annotations to existing Instance Manager (IM) pods. However, during IM upgrades or restarts, the IM controller does not insert such additional annotation into the pod spec, and the secret reconciliation in Kubernetes is not triggered—since there is no secret-related event. @NRCan-LGariepy , could you help confirm whether the role annotation will be reattached if we insert any trivial content into the credential secret, thus triggering a secret update event that would reattach the annotation to the IM pods?

**Comment by COLDTURNIP** [comment]
Reproduce steps:

1. Create a credential secret for the default backup target
2. Insert `AWS_IAM_ROLE_ARN: YXJuOmF3czppYW06OjEyMzQ1Njc4OTAxMjpyb2xlL215LXJvbGUK` into the credential secrete
3. Confirm that all instance manager pods are annotated with `iam.amazonaws.com/role: arn:aws:iam::123456789012:role/my-role`
4. Delete the instance manager pod, and the IM controller will recreate the pod.
5. Check the IM pod annotation, the IAM annotation `iam.amazonaws.com/role` is not attached.

**Comment by derekbit** [comment]
Thanks @COLDTURNIP for the steps to reproduce. Assigned to @mantissahz. Let's fix it in v1.11.0. Thanks.

**Comment by innobead** [comment]
BTW, I noticed the current implementation is only for the default backup target. Consider supporting multiple targets in the future in another ticket. 

[/controller/kubernetes_secret_controller.go#L165-L197](https://github.com/longhorn/longhorn-manager/blob/2d7b4be0bb58d2b531e16a75e23464cc605d9c76/controller/kubernetes_secret_controller.go#L165-L197)
```go
func (ks *KubernetesSecretController) reconcileSecret(namespace, secretName string) error {
	// Get default backup target
	backupTarget, err := ks.ds.GetBackupTargetRO(types.DefaultBackupTargetName)
	if err != nil {
		if !apierrors.IsNotFound(err) {
			return err
		}
		ks.logger.Warnf("Failed to find the %s backup target", types.DefaultBackupTargetName)
		return nil
	}

	backupType, err := util.CheckBackupType(backupTarget.Spec.BackupTargetURL)
	if err != nil || !types.BackupStoreRequireCredential(backupType) || backupTarget.Spec.CredentialSecret != secretName {
		// We only focus on backup target S3 or CIFS and the credential secret setting matches to the current secret name
		return nil
	}

	secret, err := ks.ds.GetSecretRO(namespace, secretName)
	if err != nil && !apierrors.IsNotFound(err) {
		return err
	}
	awsIAMRoleArn := ""
	if secret != nil {
		awsIAMRoleArn = string(secret.Data[types.AWSIAMRoleArn])
	}

	// Annotates AWS IAM role arn to the manager as well as the replica instance managers
	if err := ks.annotateAWSIAMRoleArn(awsIAMRoleArn); err != nil {
		return err
	}
	// Trigger backup_target_controller once the credential secret changes
	return ks.triggerSyncBackupTarget(backupTarget)
}
```

**Comment by yangchiu** [comment]
Verification passed on master-head (longhorn-manager 8906c76) by following [the test steps](https://github.com/longhorn/longhorn/issues/9923#issuecomment-3520345067).
